File: 백준/Gold/13244. Tree/Tree.cc
#include <bits/stdc++.h>

using namespace std;

int t, n, m, a, b, visited[1004], cnt;
vector<int> v[1004];

void dfs(int here) {
    visited[here] = 1;
    for(int there : v[here]) {
        if(!visited[there]) dfs(there);
    }
    return;
}

int main() {
    cin >> t;
    while(t--) {
        cin >> n >> m;
        
        fill(visited, visited + 1004, 0);
        cnt = 0;
        
        for(int i = 0; i < m; i++) {
            cin >> a >> b;
            v[a].push_back(b);
            v[b].push_back(a);
        }
        for(int i = 1; i <= n; i++) {
            if(!visited[i]) dfs(i), cnt++;
        }
        if(m == n - 1 && cnt == 1) {
            cout << "tree" << '\n';
        } else cout << "graph" << '\n';
    }
    
    return 0;
}

File: 백준/Gold/13244. Tree/Tree.swift
import Foundation
// (1) dfs 한번으로 되는가?
// (2) e = v - 1
func main() {
    var t = Int(readLine()!)!

    while t > 0 {
        let n = Int(readLine()!)!
        let m = Int(readLine()!)!
        var adj = Array(repeating: [Int](), count: n)
        var visited = Array(repeating: false, count: n)
        var cnt = 0

        func dfs(_ here: Int) {
            visited[here] = true

            for next in adj[here] {
                if !visited[next] {
                    dfs(next)
                }
            }
        }

        for _ in 0..<m {
            let edge = readLine()!.split(separator: " ").map { Int($0)! }
            let (a, b) = (edge[0] - 1, edge[1] - 1)

            adj[a].append(b)
            adj[b].append(a)
        }
        
        for i in 0..<n {
            if !visited[i] {
                dfs(i)
                cnt += 1
            }
        }

        if cnt == 1 && m == n - 1 { print("tree") }
        else { print("graph") }

        t -= 1
    }
}

main()